### DISCUSSION TITLE
Lesson-11 TokenURI is OK, but metamask only shows a placeholder instead of picture

### DISCUSSION THREAD

**Discussion by Osneue** [Cyfrin/foundry-full-course-cu#1040] [Q&A] [answered]
I just git clone patrick's original code, then I deploy BasicNft and mint NFT.
I import it into my metamask, the output is like this

here is the screenshot
---
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/143680960/12597ce7-db29-4521-b4ae-12e9e92ae0ce)

I test TokenURI and imageURI seperately, everything is OK.
Why does it show a placeholder instead of the actual picture?

**Comment by Niferu** [comment]
Hi @Osneue 

Try same solution as [here](https://github.com/Cyfrin/foundry-full-course-f23/discussions/1054)

**Reply by Osneue** [reply]
Thanks for your reply, but I think my problem is a little different from his, and I don't think it's caused by an incorrect URI.

I cloned patrick's repository, here is the link https://github.com/Cyfrin/foundry-nft-f23

and I tested the tokenURI and imageURI separately in the browser, and they were both normal.

I just `make anvil`, `make deploy` and `make mint`.

the tokenURI: ipfs://bafybeig37ioir76s7mg5oobetncojcm3c3hxasyd4rvid4jqhy4gkaheg4/?filename=0-PUG.json
the imageURI: https://ipfs.io/ipfs/QmSsYRx3LpDAb1GZQm7zZ1AuHZjfbPkD6J7s9r41xu1mf8?filename=pug.png

these are both normal in Brave, if I remove `?filename=0-PUG.json` and `filename=pug.png`, the problem still exists.

I'm really confused

**Reply by Niferu** [reply]
Hi @Osneue 

I have gave you complete guide what works for me. If you remove `?filename=pug.png` it wont solve the issue as you need to also post this new image to ipfs yourself then use its uri to mint NFT. You can do so by using pinata or nft.storage for example.

1. Create some image
2. Post this image to ipfs using Pinata or NFT.storage
3. Get its unique metadata uri from Pinata/NFT.storage
4. Use that metadata uri in format, which I have already showed you while minting NFT
5. Call tokenURI() function in your mother nft contract to check if metadata URI you have used lead to proper metadata on browser
6. Copy link from uri from point 5 and check in browser if it is working

If you do all the above there is no way it will not work as I was creating few NFT's recently and all are working for me.

**Reply by Osneue** [reply]
Hi @Neftyr

I followed your steps exactly

1. I post my image to ipfs using Pinata
2. the unique metadata uri from Pinata
https://white-neat-vole-513.mypinata.cloud/ipfs/QmYEZWcu13ZivPR9Z5UJwkmm22yzuN8xYeTD9p3mNBKq8Q
```
{
    "name": "PUG",
    "description": "An adorable PUG pup! (sss)",
    "image": "https://white-neat-vole-513.mypinata.cloud/ipfs/QmXoZ4EpXEo8d2p4Yi3Ss7Ro14RtwQMsHU1f59qP8deNzi",
    "attributes": [
        {
            "trait_type": "cuteness",
            "value": 100
        }
    ]
}
```
and imageURI is https://white-neat-vole-513.mypinata.cloud/ipfs/QmXoZ4EpXEo8d2p4Yi3Ss7Ro14RtwQMsHU1f59qP8deNzi
3. Use that metadata uri, you can see I use the link from Pinata
```
contract MintBasicNft is Script {
    string public constant PUG_URI =
        "https://white-neat-vole-513.mypinata.cloud/ipfs/QmYEZWcu13ZivPR9Z5UJwkmm22yzuN8xYeTD9p3mNBKq8Q";
    uint256 deployerKey;

    function run() external {
        address mostRecentlyDeployedBasicNft = DevOpsTools
            .get_most_recent_deployment("BasicNft", block.chainid);
        mintNftOnContract(mostRecentlyDeployedBasicNft);
    }

    function mintNftOnContract(address basicNftAddress) public {
        vm.startBroadcast();
        BasicNft(basicNftAddress).mintNft(PUG_URI);
        vm.stopBroadcast();
    }
}
```
```
function mintNft(string memory tokenUri) public {
        s_tokenIdToUri[s_tokenCounter] = tokenUri;
        _safeMint(msg.sender, s_tokenCounter);
        s_tokenCounter = s_tokenCounter + 1;
    }

function tokenURI(
        uint256 tokenId
    ) public view override returns (string memory) {
        if (!_exists(tokenId)) {
            revert BasicNft__TokenUriNotFound();
        }
        return s_tokenIdToUri[tokenId];
    }
```
4. I check the link in browser, it's OK
but this is what my metamask looks like
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/143680960/7a119591-317d-4a73-9ec5-432832ee8d97)
you can see the 'description' matches the json I provided, so the replacement was successful.
and image is OK in browser, 

I really don't know why. Could you provide me with a minimal successful example code? I can learn from it.

**Reply by Niferu** [reply, marked as answer]
Hi @Osneue 

Thats really strange indeed, can you please check if you see your NFT image on OpenSea? (Just remember to change network to testnet). Please also make sure you are using correct mother NFT contract while importing your NFT to metamask.

You can also give me your contract address, I will check it on etherscan.

**Reply by Osneue** [reply]
Hi @Nlferu

It's OK on OpeanSea! https://testnets.opensea.io/assets/sepolia/0xd067c7135bf5c168f6e9eb362cea4ea59a6cb821/0/
but it still looks like this in my metamask wallet
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/143680960/95a7e935-5c1f-496a-9418-138effd5f38c)
Although I don't know why it can't work on metamask, at least I know there's nothing wrong with my contract 😃
Thanks a lot, @Nlferu

**Reply by Niferu** [reply]
Hi @Osneue 

Yea thats what I thought. try one more thing. Go to meta mask, go to settings --> advanced --> clear meta mask logs.
Next go to security and privacy then enable autoDetectNft and openSea api if you do not have those.

**Reply by Osneue** [reply]
Hi, @Nlferu

I follow your suggestions.
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/143680960/4c578d3b-e316-498c-8372-b342131764c2)
and
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/143680960/e16e4826-1b8b-4fe3-90dd-2e6c843586d6)

I reimport my nft. Nothing has changed.
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/143680960/ba2affb3-f037-4269-b30b-ba1a590012e8)

**Reply by Niferu** [reply]
Hi @Osneue 
Thats very strange as it worked [here](https://github.com/Cyfrin/foundry-full-course-f23/discussions/1054)

Anyway if you can see those on OpenSea you can be sure NFT has been created correctly.
